### PR TITLE
Implement a warning that detects Swift compile units that were compiled with

### PR DESCRIPTION
**Pull request by adrian-prantl**
Implement a warning that detects Swift compile units that were compiled with a different Swift compiler than the one that is embedded in LLDB.
    
rdar://82982242

### CODE REVIEW

File: lldb/test/Shell/Swift/ToolchainMismatch.test [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# REQUIRES: swift
+# Tests that error messages from deserializing Swift modules are
```
Comment (kastiglione) [inline lldb/test/Shell/Swift/ToolchainMismatch.test]: is this message accurate? It seems to be cut short and also this doesn't seem to be related to deserializing Swift modules.

File: lldb/test/Shell/Swift/ToolchainMismatch.test [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+# RUN: rm -rf %t && mkdir %t && cd %t
+# RUN: %target-swiftc -g \
+# RUN:          -module-cache-path %t/cache %S/Inputs/main.swift \
+# RUN:          -module-name main -o a.ll -emit-ir
+# RUN: sed -i -e 's/producer: "Swift [^"]*"/producer: "Swift (swiftlang-9999.8.7.6)"/g' a.ll
```
Comment (kastiglione) [inline lldb/test/Shell/Swift/ToolchainMismatch.test]: cool

File: lldb/source/Target/TargetProperties.td [hunk 3]
```diff
@@ -241,6 +241,9 @@ let Definition = "process" in {
   def WarningUnsupportedLanguage: Property<"unsupported-language-warnings", "Boolean">,
     DefaultTrue,
     Desc<"If true, warn when stopped in code that is written in a source language that LLDB does not support.">;
+  def WarningToolchainMismatch: Property<"toolchain-mismatch-warnings", "Boolean">,
```
Comment (kastiglione) [inline lldb/source/Target/TargetProperties.td:244]: Which scenarios would people want this off?

**Review by kastiglione — APPROVED** [review]

File: lldb/source/Target/TargetProperties.td (hunk 3 above)
Comment (adrian-prantl) [reply lldb/source/Target/TargetProperties.td:244]: Any warning that potentially prints on every frame should be switchable. More specifically, I can see how this warning gets annoying when you use LLDB to debug or symbolicate crashes of previously built production software where you just don't have access to a historical matching LLDB.

File: lldb/test/Shell/Swift/ToolchainMismatch.test (hunk 1 above)
Comment (adrian-prantl) [reply lldb/test/Shell/Swift/ToolchainMismatch.test]: copy&pasto

**Comment by adrian-prantl** [comment]
Also added a sanity check to the test.

**Comment by adrian-prantl** [comment]
@JDevlieghere Do you think it's reasonable to expect the swiftlang tag to match or should we just match the three most significant components of the version number?

**Comment by kastiglione** [comment]
One thing to add: this change may be complimentary, or redundant, to a change in swiftc where it will begin to embed the sdk into swiftmodules. When that happens, deserializing such swiftmodules will return a status of `SDKMismatch`. See https://github.com/apple/swift-driver/pull/838 and https://github.com/apple/swift/pull/37768.

**Comment by adrian-prantl** [comment]
After discussing this with @JDevlieghere I changed the check to leave off the build number, to reduce the risk of benign mismatches triggering false positives. Due to procedural reasons the compiler and debugger may be legitimately a few numbers apart.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by shahmishal** [comment]
@swift-ci test

File: lldb/include/lldb/Target/Process.h [hunk 4]
```diff
@@ -93,6 +93,9 @@ class ProcessProperties : public Properties {
   void SetDetachKeepsStopped(bool keep_stopped);
   bool GetWarningsOptimization() const;
   bool GetWarningsUnsupportedLanguage() const;
+#ifdef LLDB_ENABLE_SWIFT
+  bool GetWarningsToolchainMismatch() const;
```
Comment (JDevlieghere) [inline lldb/include/lldb/Target/Process.h:97]: As a follow up patch, can we get rid of this function and the `PrintWarning.*` ones below by moving them into the language plugin?

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp [hunk 5]
```diff
@@ -609,52 +609,45 @@ bool DWARFUnit::DW_AT_decl_file_attributes_are_invalid() {
...
-    const uint32_t major_version = GetProducerVersionMajor();
-    return major_version > 425 ||
-           (major_version == 425 && GetProducerVersionUpdate() >= 13);
-  }
-  return true; // Assume all other compilers didn't have incorrect ObjC bitfield
-               // info
+  if (GetProducer() == eProducerClang)
+    return GetProducerVersion() >= llvm::VersionTuple(425, 0, 13);
+  // Assume all other compilers didn't have incorrect ObjC bitfield info.
+  return true;
```
Comment (JDevlieghere) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp:615]: Why can't this live upstream?

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp (hunk 5 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp:615]: This is a cherry-pick.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test
